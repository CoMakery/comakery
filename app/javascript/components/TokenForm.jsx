import React from 'react'
import PropTypes from 'prop-types'
import {fetch as fetchPolyfill} from 'whatwg-fetch'
import InputFieldDropdownHalfed from './styleguide/InputFieldDropdownHalfed'
import InputFieldHalfed from './styleguide/InputFieldHalfed'
import InputFieldUploadFile from './styleguide/InputFieldUploadFile'
import Button from './styleguide/Button'

class TokenForm extends React.Component {
  constructor(props) {
    super(props)

    this.errorAdd = this.errorAdd.bind(this)
    this.errorRemove = this.errorRemove.bind(this)
    this.disable = this.disable.bind(this)
    this.enable = this.enable.bind(this)
    this.disableContractFields = this.disableContractFields.bind(this)
    this.enableContractFields = this.enableContractFields.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleFieldChange = this.handleFieldChange.bind(this)
    this.handleLogoChange = this.handleLogoChange.bind(this)
    this.fetchSymbolAndDecimals = this.fetchSymbolAndDecimals.bind(this)

    this.state = {
      logoLocalUrl                      : null,
      errors                            : {},
      disabled                          : {},
      'token[coin_type]'                : this.props.token.coinType || Object.values(this.props.coinTypes)[0],
      'token[name]'                     : this.props.token.name || '',
      'token[symbol]'                   : this.props.token.symbol || '',
      'token[contract_address]'         : this.props.token.contractAddress || '',
      'token[ethereum_contract_address]': this.props.token.ethereumContractAddress || '',
      'token[decimal_places]'           : (!this.props.token.decimalPlaces && this.props.token.decimalPlaces !== 0 ? '' : this.props.token.decimalPlaces),
      'token[blockchain_network]'       : this.props.token.blockchainNetwork || Object.values(this.props.blockchainNetworks)[0],
      'token[ethereum_network]'         : this.props.token.ethereumNetwork || Object.values(this.props.ethereumNetworks)[0]
    }
  }

  componentDidUpdate() {
    console.log(this.state)
  }

  errorAdd(n, e) {
    this.setState({
      errors: Object.assign({}, this.state.errors, {[n]: e})
    })
  }

  errorRemove(n) {
    let e = this.state.errors
    delete e[n]
    this.setState({
      errors: e
    })
  }

  disable(a) {
    let d = this.state.disabled
    a.forEach(n => d = Object.assign({}, d, {[n]: true}))
    this.setState({
      disabled: d
    })
  }

  enable(a) {
    let d = this.state.disabled
    a.forEach(n => delete d[n])
    this.setState({
      disabled: d
    })
  }

  disableContractFields() {
    this.disable([
      'token[coin_type]',
      'token[blockchain_network]',
      'token[ethereum_network]',
      'token[contract_address]',
      'token[ethereum_contract_address]'
    ])
  }

  enableContractFields() {
    this.enable([
      'token[coin_type]',
      'token[blockchain_network]',
      'token[ethereum_network]',
      'token[contract_address]',
      'token[ethereum_contract_address]'
    ])
  }

  handleFieldChange(event) {
    this.setState({ [event.target.name]: event.target.value })

    if (!event.target.checkValidity()) {
      this.errorAdd(event.target.name, 'invalid value')
      return
    } else {
      this.errorRemove(event.target.name)
    }

    if (event.target.value === '') {
      return
    }

    switch (event.target.name) {
      case 'token[contract_address]':
        this.fetchSymbolAndDecimals(event.target.value, this.state['token[blockchain_network]'])
        break

      case 'token[blockchain_network]':
        this.fetchSymbolAndDecimals(this.state['token[contract_address]'], event.target.value)
        break

      case 'token[ethereum_contract_address]':
        this.fetchSymbolAndDecimals(event.target.value, this.state['token[ethereum_network]'])
        break

      case 'token[ethereum_network]':
        this.fetchSymbolAndDecimals(this.state['token[ethereum_contract_address]'], event.target.value)
        break

      case 'token[coin_type]':
        this.setState({
          'token[symbol]'                   : '',
          'token[decimal_places]'           : '',
          'token[contract_address]'         : '',
          'token[ethereum_contract_address]': ''
        })
        this.enableContractFields()
        break

      default:
        return
    }
  }

  handleLogoChange(event) {
    if (this.state.logoLocalUrl != null) {
      URL.revokeObjectURL(this.state.logoLocalUrl)
    }
    this.setState({
      logoLocalUrl: URL.createObjectURL(event.target.files[0])
    })
  }

  fetchSymbolAndDecimals(address, network) {
    if (address === '' || network === '') {
      return
    }

    this.setState({
      'token[symbol]'        : '',
      'token[decimal_places]': ''
    })
    this.disableContractFields()

    fetchPolyfill('/tokens/fetch_contract_details', {
      credentials: 'same-origin',
      method     : 'POST',
      body       : JSON.stringify({'address': address, 'network': network, 'authenticity_token': this.props.csrfToken}),
      headers    : {
        'Accept'      : 'application/json',
        'Content-Type': 'application/json'
      }
    }).then(response => {
      if (response.status === 200) {
        return response.json()
      } else {
        this.enableContractFields()
        throw Error(response.text())
      }
    }).then(data => {
      let symbol = data.symbol
      let decimals = data.decimals
      if (symbol && decimals) {
        this.setState({
          'token[symbol]'        : symbol,
          'token[decimal_places]': decimals.toString()
        })
      }
      this.enableContractFields()
    })
  }

  handleSubmit(event) {
    event.preventDefault()

    this.disable(['token[submit]'])

    if (!event.target.checkValidity()) {
      this.enable(['token[submit]'])
      return
    }

    const formData = new FormData(event.target)

    fetchPolyfill(this.props.formUrl, {
      credentials: 'same-origin',
      method     : this.props.formAction,
      body       : formData,
      headers    : {
        'X-Key-Inflection': 'snake'
      }
    }).then(response => {
      if (response.status === 200) {
        window.location = this.props.urlOnSuccess
      } else {
        response.json().then(data => {
          this.setState({
            errors: data.errors
          })
          this.enable(['token[submit]'])
        })
      }
    })
  }

  render() {
    return (
      <React.Fragment>
        <form onSubmit={this.handleSubmit}>

          <InputFieldDropdownHalfed
            title="payment type"
            required
            name="token[coin_type]"
            value={this.state['token[coin_type]']}
            errorText={this.state.errors['token[coin_type]']}
            disabled={this.state.disabled['token[coin_type]']}
            eventHandler={this.handleFieldChange}
            selectEntries={Object.entries(this.props.coinTypes)}
            symbolLimit={0}
          />

          <InputFieldHalfed
            title="token name"
            required
            name="token[name]"
            value={this.state['token[name]']}
            errorText={this.state.errors['token[name]']}
            placeholder="Bitcoin"
            eventHandler={this.handleFieldChange}
            symbolLimit={0}
          />

          {this.state['token[coin_type]'].match(/qrc20|erc20/) &&
            <InputFieldHalfed
              title="token symbol"
              required
              name="token[symbol]"
              value={this.state['token[symbol]']}
              errorText={this.state.errors['token[symbol]']}
              placeholder="..."
              readOnly
              eventHandler={this.handleFieldChange}
              symbolLimit={0}
            />
          }

          {this.state['token[coin_type]'].match(/qrc20|erc20/) &&
            <InputFieldHalfed
              title="decimal places"
              required
              name="token[decimal_places]"
              value={this.state['token[decimal_places]']}
              errorText={this.state.errors['token[decimal_places]']}
              placeholder="..."
              pattern="\d{1-2}"
              readOnly
              eventHandler={this.handleFieldChange}
              symbolLimit={0}
            />
          }

          {this.state['token[coin_type]'] === 'qrc20' &&
            <InputFieldHalfed
              title="contract address"
              required
              name="token[contract_address]"
              value={this.state['token[contract_address]']}
              errorText={this.state.errors['token[contract_address]']}
              readOnly={this.state.disabled['token[contract_address]']}
              placeholder="2c754a7b03927a5a30ca2e7c98a8fdfaf17d11fc"
              pattern="[a-fA-F0-9]{40}"
              eventHandler={this.handleFieldChange}
              symbolLimit={0}
            />
          }

          {this.state['token[coin_type]'] === 'erc20' &&
            <InputFieldHalfed
              title="contract address"
              required
              name="token[ethereum_contract_address]"
              value={this.state['token[ethereum_contract_address]']}
              errorText={this.state.errors['token[ethereum_contract_address]']}
              readOnly={this.state.disabled['token[ethereum_contract_address]']}
              placeholder="0x6c6ee5e31d828de241282b9606c8e98ea48526e2"
              pattern="0x[a-fA-F0-9]{40}"
              eventHandler={this.handleFieldChange}
              symbolLimit={0}
            />
          }

          {this.state['token[coin_type]'] === 'qrc20' &&
            <InputFieldDropdownHalfed
              title="blockchain network"
              required
              name="token[blockchain_network]"
              value={this.state['token[blockchain_network]']}
              errorText={this.state.errors['token[blockchain_network]']}
              disabled={this.state.disabled['token[blockchain_network]']}
              eventHandler={this.handleFieldChange}
              selectEntries={Object.entries(this.props.blockchainNetworks)}
            />
          }

          {this.state['token[coin_type]'].match(/eth|erc20/) &&
            <InputFieldDropdownHalfed
              title="blockchain network"
              required
              name="token[ethereum_network]"
              value={this.state['token[ethereum_network]']}
              errorText={this.state.errors['token[ethereum_network]']}
              disabled={this.state.disabled['token[ethereum_network]']}
              eventHandler={this.handleFieldChange}
              selectEntries={Object.entries(this.props.ethereumNetworks)}
            />
          }

          <InputFieldUploadFile
            title="token logo"
            required
            name="token[logo_image]"
            errorText={this.state.errors['token[logo_image]']}
            imgPreviewUrl={this.props.token.logoUrl}
          />

          <input
            type="hidden"
            name="authenticity_token"
            value={this.props.csrfToken}
            readOnly
          />

          <Button
            type="submit"
            value="save"
            disabled={this.state.disabled.submit}
          />
        </form>
      </React.Fragment>
    )
  }
}

TokenForm.propTypes = {
  token             : PropTypes.object.isRequired,
  coinTypes         : PropTypes.object.isRequired,
  ethereumNetworks  : PropTypes.object.isRequired,
  blockchainNetworks: PropTypes.object.isRequired,
  formUrl           : PropTypes.string.isRequired,
  formAction        : PropTypes.string.isRequired,
  urlOnSuccess      : PropTypes.string.isRequired,
  csrfToken         : PropTypes.string.isRequired
}
TokenForm.defaultProps = {
  token             : {},
  coinTypes         : {},
  ethereumNetworks  : {},
  blockchainNetworks: {},
  formUrl           : '',
  formAction        : '',
  urlOnSuccess      : '',
  csrfToken         : ''
}
export default TokenForm
