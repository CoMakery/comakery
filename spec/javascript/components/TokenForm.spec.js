import React from 'react'
import { mount } from 'enzyme'
import TokenForm from 'components/TokenForm'

describe('TokenForm', () => {
  beforeEach(() => {
    fetch.resetMocks()
  })

  it('renders correctly without props', () => {
    const wrapper = mount(<TokenForm />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('Layout[title="Create a New Token"]')).toBe(true)
    expect(wrapper.exists('.token-form')).toBe(true)
    expect(wrapper.exists('ButtonBorder[value="cancel"]')).toBe(true)
    expect(wrapper.exists('ButtonBorder[value="create"]')).toBe(true)
    expect(wrapper.exists('Button[value="create & close"]')).toBe(true)
    expect(wrapper.exists('.token-form--message')).toBe(false)
    expect(wrapper.exists('.token-form--form')).toBe(true)
    expect(wrapper.exists('#token-form--form')).toBe(true)

    expect(wrapper.exists(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDropdownHalfed[title="visibility"][required][name="token[unlisted]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldUploadFile[title="token logo"][required][name="token[logo_image]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'input[type="hidden"][name="authenticity_token"]'
    )).toBe(true)
  })

  it('renders correctly with erc token', () => {
    const token = {
      'id'                     : 0,
      'name'                   : 'ERC-TEST',
      'coinType'               : 'erc20',
      'blockchainNetwork'      : 'main',
      'contractAddress'        : '0x00',
      'symbol'                 : 'ERCT',
      'decimalPlaces'          : 2,
      'logoUrl'                : '/ERCT.png'
    }
    const wrapper = mount(<TokenForm token={token} />)

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    ).props().value).toBe('erc20')

    expect(wrapper.find(
      'InputFieldHalfed[title="token name"][required][name="token[name]"][placeholder="Bitcoin"]'
    ).props().value).toBe('ERC-TEST')

    expect(wrapper.find(
      'InputFieldUploadFile[title="token logo"][required][name="token[logo_image]"]'
    ).props().imgPreviewUrl).toBe('/ERCT.png')

    expect(wrapper.find(
      'InputFieldHalfed[title="token symbol"][required][readOnly][name="token[symbol]"][placeholder="..."]'
    ).props().value).toBe('ERCT')

    expect(wrapper.find(
      'InputFieldHalfed[title="decimal places"][required][readOnly][name="token[decimal_places]"][placeholder="..."]'
    ).props().value).toBe('2')

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="blockchain network"][required][name="token[blockchain_network]"]'
    ).props().value).toBe('main')
  })

  it('renders correctly with qrc token', () => {
    const token = {
      'id'                     : 1,
      'name'                   : 'QRC-TEST',
      'coinType'               : 'qrc20',
      'blockchainNetwork'      : 'test',
      'contractAddress'        : '0',
      'symbol'                 : 'QRCT',
      'decimalPlaces'          : 0,
      'logoUrl'                : '/QRCT.png'
    }
    const wrapper = mount(<TokenForm token={token} />)

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    ).props().value).toBe('qrc20')

    expect(wrapper.find(
      'InputFieldHalfed[title="token name"][required][name="token[name]"][placeholder="Bitcoin"]'
    ).props().value).toBe('QRC-TEST')

    expect(wrapper.find(
      'InputFieldUploadFile[title="token logo"][required][name="token[logo_image]"]'
    ).props().imgPreviewUrl).toBe('/QRCT.png')

    expect(wrapper.find(
      'InputFieldHalfed[title="contract address"][required][name="token[contract_address]"][placeholder="2c754a7b03927a5a30ca2e7c98a8fdfaf17d11fc"][pattern="[a-fA-F0-9]{40}"]'
    ).props().value).toBe('0')

    expect(wrapper.find(
      'InputFieldHalfed[title="token symbol"][required][readOnly][name="token[symbol]"][placeholder="..."]'
    ).props().value).toBe('QRCT')

    expect(wrapper.find(
      'InputFieldHalfed[title="decimal places"][required][readOnly][name="token[decimal_places]"][placeholder="..."]'
    ).props().value).toBe('0')

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="blockchain network"][required][name="token[blockchain_network]"]'
    ).props().value).toBe('test')
  })

  it('renders correctly with eth token', () => {
    const token = {
      'id'                     : 1,
      'name'                   : 'ETH-TEST',
      'coinType'               : 'eth',
      'blockchainNetwork'      : 'test',
      'contractAddress'        : null,
      'symbol'                 : null,
      'decimalPlaces'          : null,
      'logoUrl'                : '/ETH.png'
    }
    const wrapper = mount(<TokenForm token={token} />)

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    ).props().value).toBe('eth')

    expect(wrapper.find(
      'InputFieldUploadFile[title="token logo"][required][name="token[logo_image]"]'
    ).props().imgPreviewUrl).toBe('/ETH.png')

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="blockchain network"][required][name="token[blockchain_network]"]'
    ).props().value).toBe('test')
  })

  it('renders correctly with coinTypes', () => {
    const coinTypes = {
      'ERC20': 'erc20',
      'QRC20': 'qrc20',
      'ETH'  : 'eth'
    }
    const wrapper = mount(<TokenForm coinTypes={coinTypes} />)

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    ).props().value).toBe('erc20')

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="payment type"][required][name="token[coin_type]"]'
    ).props().selectEntries).toEqual(Object.entries(coinTypes))
  })

  it('renders correctly with blockchainNetworks', () => {
    const token = {
      'id'                     : 1,
      'name'                   : 'QRC-TEST',
      'coinType'               : 'qrc20',
      'blockchainNetwork'      : null,
      'contractAddress'        : '0',
      'symbol'                 : 'QRCT',
      'decimalPlaces'          : 0,
      'logoUrl'                : '/QRCT.png'
    }
    const blockchainNetworks = {
      'main QTUM Network': 'qtum_mainnet',
      'test QTUM Network': 'qtum_testnet'
    }
    const wrapper = mount(<TokenForm token={token} blockchainNetworks={blockchainNetworks} />)

    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="blockchain network"][required][name="token[blockchain_network]"]'
    ).props().value).toBe('qtum_mainnet')
    expect(wrapper.find(
      'InputFieldDropdownHalfed[title="blockchain network"][required][name="token[blockchain_network]"]'
    ).props().selectEntries).toEqual(Object.entries(blockchainNetworks))
  })

  it('renders correctly with csrfToken', () => {
    const wrapper = mount(<TokenForm csrfToken='test' />)

    expect(wrapper.find(
      'input[type="hidden"][name="authenticity_token"]'
    ).props().value).toBe('test')
  })

  it('uses formUrl', () => {
    const wrapper = mount(<TokenForm formUrl='/test' />)

    expect(wrapper.state('formUrl')).toBe('/test')
  })

  it('uses formAction', () => {
    const wrapper = mount(<TokenForm formAction='PUT' />)

    expect(wrapper.state('formAction')).toBe('PUT')
  })

  it('uses urlOnSuccess', () => {
    const wrapper = mount(<TokenForm urlOnSuccess='/test' />)

    expect(wrapper.props().urlOnSuccess).toBe('/test')
  })

  it('displays flash messages', () => {
    const wrapper = mount(<TokenForm />)

    wrapper.setState({
      flashMessages: [
        {
          'severity': 'notice',
          'text'    : 'notice text'
        },
        {
          'severity': 'warning',
          'text'    : 'warning text'
        },
        {
          'severity': 'error',
          'text'    : 'error text'
        }
      ]
    })

    wrapper.update()

    expect(wrapper.exists('Flash')).toBe(true)
  })

  it('displays errors', () => {
    const wrapper = mount(<TokenForm />)

    wrapper.setState({
      errors: {
        'token[coin_type]' : 'coin_type error',
        'token[logo_image]': 'logo_image error'
      }
    })

    wrapper.update()

    expect(wrapper.exists(
      'InputFieldDropdownHalfed[errorText="coin_type error"][title="payment type"][required][name="token[coin_type]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldUploadFile[errorText="logo_image error"][title="token logo"][required][name="token[logo_image]"]'
    )).toBe(true)
  })
})
