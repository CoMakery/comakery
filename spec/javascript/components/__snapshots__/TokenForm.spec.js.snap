// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenForm renders correctly without props 1`] = `
<TokenForm
  blockchainNetworks={
    Object {
      "default": "_",
    }
  }
  coinTypes={
    Object {
      "default": "_",
    }
  }
  csrfToken="00"
  ethereumNetworks={
    Object {
      "default": "_",
    }
  }
  formAction="POST"
  formUrl="/"
  token={
    Object {
      "default": "_",
    }
  }
  urlOnSuccess="/"
>
  <Layout
    className="token-form"
    hasBackButton={true}
    sidebar={null}
    subfooter={
      <React.Fragment>
        <ButtonBorder
          className=""
          onClick={[Function]}
          value="cancel"
        />
        <ButtonBorder
          className=""
          form="token-form--form"
          onClick={[Function]}
          type="submit"
          value="create"
        />
        <Button
          className=""
          disabled={false}
          form="token-form--form"
          onClick={[Function]}
          type="submit"
          value="create & close"
        />
      </React.Fragment>
    }
    title="Create a New Token"
  >
    <div
      className="layout token-form"
    >
      <div
        className="layout--back-button"
      >
        <Icon
          className=""
          name="iconCloseCopy.svg"
          onClick={[Function]}
        >
          <img
            className="icon icon__iconCloseCopy-svg"
            onClick={[Function]}
            src="test-file-stub"
          />
        </Icon>
      </div>
      <div
        className="layout--content"
      >
        <div
          className="layout--content--title"
        >
          Create a New Token
        </div>
        <hr
          className="layout--content--hr"
        />
        <div
          className="layout--content--wrapper"
        >
          <div
            className="layout--content--content"
          >
            <form
              className="token-form--form"
              id="token-form--form"
              onSubmit={[Function]}
            >
              <InputFieldDropdownHalfed
                eventHandler={[Function]}
                name="token[coin_type]"
                required={true}
                selectEntries={
                  Array [
                    Array [
                      "default",
                      "_",
                    ],
                  ]
                }
                symbolLimit={0}
                title="payment type"
                value="_"
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__dropdown--halfed"
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  name="token[coin_type]"
                  pattern=".*"
                  placeholder="Please enter value"
                  readOnly={false}
                  required={true}
                  selectEntries={
                    Array [
                      Array [
                        "default",
                        "_",
                      ],
                    ]
                  }
                  symbolLimit={0}
                  title="payment type"
                  type="select"
                  value="_"
                >
                  <div
                    className="input-field input-field__dropdown--halfed"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        payment type
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <select
                        className="input-field--content__select"
                        disabled={false}
                        name="token[coin_type]"
                        onChange={[Function]}
                        required={true}
                        value="_"
                      >
                        <option
                          key="_"
                          value="_"
                        >
                          default
                        </option>
                      </select>
                    </div>
                  </div>
                </InputField>
              </InputFieldDropdownHalfed>
              <InputFieldHalfed
                eventHandler={[Function]}
                name="token[name]"
                placeholder="Bitcoin"
                required={true}
                symbolLimit={0}
                title="token name"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__halfed"
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  name="token[name]"
                  pattern=".*"
                  placeholder="Bitcoin"
                  readOnly={false}
                  required={true}
                  selectEntries={Array []}
                  symbolLimit={0}
                  title="token name"
                  type="text"
                  value=""
                >
                  <div
                    className="input-field input-field__halfed"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        token name
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <input
                        className="input-field--content__text"
                        name="token[name]"
                        onChange={[Function]}
                        pattern=".*"
                        placeholder="Bitcoin"
                        readOnly={false}
                        required={true}
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldHalfed>
              <InputFieldUploadFile
                name="token[logo_image]"
                required={true}
                title="token logo"
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__upload-file"
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  name="token[logo_image]"
                  pattern=".*"
                  placeholder="Please enter value"
                  readOnly={false}
                  required={true}
                  selectEntries={Array []}
                  symbolLimit={0}
                  title="token logo"
                  type="file"
                  value=""
                >
                  <div
                    className="input-field input-field__upload-file"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        token logo
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <input
                        className="input-field--content__file"
                        name="token[logo_image]"
                        onChange={[Function]}
                        required={true}
                        type="file"
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldUploadFile>
              <input
                name="authenticity_token"
                readOnly={true}
                type="hidden"
                value="00"
              />
            </form>
          </div>
        </div>
      </div>
      <div
        className="layout--subfooter"
      >
        <div
          className="layout--subfooter--buttons"
        >
          <ButtonBorder
            className=""
            onClick={[Function]}
            value="cancel"
          >
            <Button
              className="button__border"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="cancel"
            >
              <input
                className="button button__border"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="cancel"
              />
            </Button>
          </ButtonBorder>
          <ButtonBorder
            className=""
            form="token-form--form"
            onClick={[Function]}
            type="submit"
            value="create"
          >
            <Button
              className="button__border"
              disabled={false}
              form="token-form--form"
              onClick={[Function]}
              type="submit"
              value="create"
            >
              <input
                className="button button__border"
                disabled={false}
                form="token-form--form"
                onClick={[Function]}
                type="submit"
                value="create"
              />
            </Button>
          </ButtonBorder>
          <Button
            className=""
            disabled={false}
            form="token-form--form"
            onClick={[Function]}
            type="submit"
            value="create & close"
          >
            <input
              className="button"
              disabled={false}
              form="token-form--form"
              onClick={[Function]}
              type="submit"
              value="create & close"
            />
          </Button>
        </div>
      </div>
    </div>
  </Layout>
</TokenForm>
`;
